src/main/java/eu/europeana/metis/sandbox/service/metrics/MetricsServiceImpl.java

```diff
@@ -0,0 +1,226 @@
+      LOGGER.debug("Report processed:{} total:{}", report.getProcessedRecords(), report.getTotalRecords());
+      saveMetricsProgressDataset(datasetId, report);
+      saveMetricsProgressByStep(datasetId, report);
+    } catch (RuntimeException metricsException) {
+      LOGGER.error("Something went wrong during processing metrics", metricsException);
+    } finally {
+      lock.unlock();
+      LOGGER.debug("Process metrics: {} lock, Unlocked", datasetId);
+    }
+  }
+
+  private void saveMetricsProgressByStep(String datasetId, ProgressInfoDto report) {
+    report.getProgressByStep().stream().forEach(
+        item -> {
+          LOGGER.debug("step:{} total:{} success:{} fail:{} warn: {}",
+              item.getStep().value(), item.getTotal(),
+              item.getSuccess(), item.getFail(), item.getWarn());
+          ProgressStepEntity progressStepEntity = findProgressStep(datasetId, item.getStep());
+          if (progressStepEntity == null) {
+            progressStepEntity = new ProgressStepEntity();
+            progressStepEntity.setDatasetId(datasetId);
+            progressStepEntity.setStep(item.getStep().value());
+          }
+          progressStepEntity.setSuccess(item.getSuccess());
+          progressStepEntity.setFail(item.getFail());
+          progressStepEntity.setWarn(item.getWarn());
+          progressStepEntity.setTotal(item.getTotal());
+          progressStepRepository.save(progressStepEntity);
+        }
+    );
+  }
+
+  private void saveMetricsProgressDataset(String datasetId, ProgressInfoDto report) {
+    ProgressDatasetEntity progressDatasetEntity = findProgressDataset(datasetId);
+    if (progressDatasetEntity == null) {
+      Optional<DatasetEntity> datasetEntity = datasetRepository.findById(Integer.valueOf(datasetId));
+      LocalDateTime startTimeStamp = datasetEntity.isPresent() ? datasetEntity.get().getCreatedDate() : LocalDateTime.now();
+      progressDatasetEntity = new ProgressDatasetEntity();
+      progressDatasetEntity.setDatasetId(datasetId);
+      progressDatasetEntity.setStartTimeStamp(startTimeStamp);
+    } else {
+      progressDatasetEntity = progressDatasetRepository.findByDatasetId(datasetId);
+    }
+    progressDatasetEntity.setProcessedRecords(report.getProcessedRecords());
+    progressDatasetEntity.setTotalRecords(report.getTotalRecords());
+    progressDatasetEntity.setStatus(report.getStatus().value());
+    progressDatasetEntity.setEndTimeStamp(LocalDateTime.now());
+    progressDatasetRepository.save(progressDatasetEntity);
+  }
+
+  private ProgressDatasetEntity findProgressDataset(String datasetId) {
+    return progressDatasetRepository.findByDatasetId(datasetId);
+  }
+
+  private ProgressStepEntity findProgressStep(String datasetId, Step step) {
+    return progressStepRepository.findByDatasetIdAndStep(datasetId, step.value());
+  }
+
+  private void getProgressDatasetMetrics(Map<String, Object> metricsMap) {
```

Comment:
This is not really a get is it? It is adding values to a passed parameter, so I would name this differently.  
That being said though, I usually try to avoid modifying/extending the passed parameters and return the result I need.  
The caller then extends(copies all) its own object with the returned results.

Reply:
Refactored.